astroid/brain/brain_dataclasses.py

```diff
@@ -235,8 +228,48 @@ def _generate_dataclass_init(assigns: list[AnnAssign]) -> str:
         if not init_var:
             assignments.append(assignment_str)
 
-    params_string = ", ".join(["self"] + params)
     assignments_string = "\n    ".join(assignments) if assignments else "pass"
+    new_params_string = ", ".join(params)
+
+    try:
+        base: ClassDef = next(next(iter(node.bases)).infer())
+        base_init: FunctionDef | None = base.locals["__init__"][0]
+    except (StopIteration, InferenceError, KeyError):
+        base_init = None
+
+    prev_pos_only = ""
+    prev_kw_only = ""
+    if base_init and base.is_dataclass:
+        # Skip the self argument
+        all_arguments = base_init.args.format_args()[6:].split(", ")
+        arguments = ", ".join(
+            i for i in all_arguments if i.split(":")[0] not in new_params_string
+        )
```

Comment:
I managed to break this assumption because "e" is contained in "ee":

This test will now fail:
```diff
diff --git a/tests/unittest_brain_dataclasses.py b/tests/unittest_brain_dataclasses.py
index a2077d1c..92a61a85 100644
--- a/tests/unittest_brain_dataclasses.py
+++ b/tests/unittest_brain_dataclasses.py
@@ -761,7 +761,7 @@ def test_kw_only_decorator() -> None:
     @dataclass(kw_only=True)
     class Foo:
         a: int
-        b: str
+        e: str
 
 
     @dataclass(kw_only=False)
@@ -776,7 +776,7 @@ def test_kw_only_decorator() -> None:
 
     @dataclass(kw_only=True)
     class Dee(Cee):
-        e: int
+        ee: int
 
 
     Foo.__init__  #@
@@ -789,31 +789,31 @@ def test_kw_only_decorator() -> None:
     foo_init: bases.UnboundMethod = next(foodef.infer())
     if PY310_PLUS:
         assert [a.name for a in foo_init.args.args] == ["self"]
-        assert [a.name for a in foo_init.args.kwonlyargs] == ["a", "b"]
+        assert [a.name for a in foo_init.args.kwonlyargs] == ["a", "e"]
     else:
-        assert [a.name for a in foo_init.args.args] == ["self", "a", "b"]
+        assert [a.name for a in foo_init.args.args] == ["self", "a", "e"]
         assert [a.name for a in foo_init.args.kwonlyargs] == []
 
     bar_init: bases.UnboundMethod = next(bardef.infer())
     if PY310_PLUS:
         assert [a.name for a in bar_init.args.args] == ["self", "c"]
-        assert [a.name for a in bar_init.args.kwonlyargs] == ["a", "b"]
+        assert [a.name for a in bar_init.args.kwonlyargs] == ["a", "e"]
     else:
-        assert [a.name for a in bar_init.args.args] == ["self", "a", "b", "c"]
+        assert [a.name for a in bar_init.args.args] == ["self", "a", "e", "c"]
         assert [a.name for a in bar_init.args.kwonlyargs] == []
 
     cee_init: bases.UnboundMethod = next(cee.infer())
     if PY310_PLUS:
         assert [a.name for a in cee_init.args.args] == ["self", "c", "d"]
-        assert [a.name for a in cee_init.args.kwonlyargs] == ["a", "b"]
+        assert [a.name for a in cee_init.args.kwonlyargs] == ["a", "e"]
     else:
-        assert [a.name for a in cee_init.args.args] == ["self", "a", "b", "c", "d"]
+        assert [a.name for a in cee_init.args.args] == ["self", "a", "e", "c", "d"]
         assert [a.name for a in cee_init.args.kwonlyargs] == []
 
     dee_init: bases.UnboundMethod = next(dee.infer())
     if PY310_PLUS:
         assert [a.name for a in dee_init.args.args] == ["self", "c", "d"]
-        assert [a.name for a in dee_init.args.kwonlyargs] == ["a", "b", "e"]
+        assert [a.name for a in dee_init.args.kwonlyargs] == ["a", "e", "e"]
     else:
-        assert [a.name for a in dee_init.args.args] == ["self", "a", "b", "c", "d", "e"]
+        assert [a.name for a in dee_init.args.args] == ["self", "a", "e", "c", "d", "ee"]
         assert [a.name for a in dee_init.args.kwonlyargs] == []
```

Reply:
😅